pkg/reconciler/apis/crdcleanup/crdcleanup_controller.go

```diff
@@ -0,0 +1,345 @@
+) (*controller, error) {
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), ControllerName)
+
+	c := &controller{
+		queue: queue,
+		getCRD: func(clusterName logicalcluster.Name, name string) (*apiextensionsv1.CustomResourceDefinition, error) {
+			return crdInformer.Lister().Cluster(clusterName).Get(name)
+		},
+		getAPIBindingsByBoundResourceUID: func(name string) ([]interface{}, error) {
+			return apiBindingInformer.Informer().GetIndexer().ByIndex(indexers.APIBindingByBoundResourceUID, name)
+		},
+		listAPIBindings: func() ([]*apisv1alpha1.APIBinding, error) {
+			return apiBindingInformer.Lister().List(labels.Everything())
+		},
+		getResourceUIDs: func(binding *apisv1alpha1.APIBinding) ([]types.UID, error) {
+			exportCluster := logicalcluster.New(binding.Spec.Reference.Workspace.Path)
+			export, err := apiExportInformer.Lister().Get(client.ToClusterAwareKey(exportCluster, binding.Spec.Reference.Workspace.ExportName))
+
+			if errors.IsNotFound(err) {
+				export, err = temporaryRemoteShardAPIExportInformer.Lister().Get(client.ToClusterAwareKey(exportCluster, binding.Spec.Reference.Workspace.ExportName))
+			}
+
+			if err != nil {
+				return nil, err
+			}
+
+			uids := []types.UID{}
+			for _, schemaName := range export.Spec.LatestResourceSchemas {
+				schema, err := apiResourceSchemaInformer.Lister().Get(client.ToClusterAwareKey(exportCluster, schemaName))
+
+				if errors.IsNotFound(err) {
+					schema, err = temporaryRemoteShardAPIResourceSchemaInformer.Lister().Get(client.ToClusterAwareKey(exportCluster, schemaName))
+				}
+
+				if err != nil {
+					return nil, err
+				}
+
+				uids = append(uids, schema.UID)
+			}
+			return uids, nil
+		},
+		deleteCRD: func(ctx context.Context, name string) error {
+			return crdClusterClient.ApiextensionsV1().CustomResourceDefinitions().Cluster(apibinding.ShadowWorkspaceName).Delete(ctx, name, metav1.DeleteOptions{})
+		},
+	}
+
+	indexers.AddIfNotPresentOrDie(
+		apiBindingInformer.Informer().GetIndexer(),
+		cache.Indexers{
+			indexers.APIBindingByBoundResourceUID: indexers.IndexAPIBindingByBoundResourceUID,
+		},
+	)
+
+	crdInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			crd, ok := obj.(*apiextensionsv1.CustomResourceDefinition)
+
+			if !ok {
+				return false
+			}
+
+			return logicalcluster.From(crd) == apibinding.ShadowWorkspaceName
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				c.enqueueCRD(obj)
+			},
+			UpdateFunc: func(_, newObj interface{}) {
+				c.enqueueCRD(newObj)
+			},
+			DeleteFunc: func(obj interface{}) {
+				c.enqueueCRD(obj)
+			},
+		},
+	})
+
+	apiBindingInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				c.enqueueFromAPIBinding(oldObj, newObj)
+			},
+			DeleteFunc: func(obj interface{}) {
+				c.enqueueFromAPIBinding(obj, obj)
```

Comment:
this is strange. The method says old and new, but now you call it was obj and obj. Would expect `enqueueFromAPIBinding(nil, obj)`.